Index: CDFFileData.cpp
===================================================================
--- CDFFileData.cpp	(revision 44417)
+++ CDFFileData.cpp	(working copy)
@@ -747,8 +747,8 @@
 		ReadNextLine(instr, str, MAXLINELENGTH); // #qc ProbeSets
 		subStr=strchr(str,'=')+1;
 		m_Header.m_NumQCProbeSets = atoi(subStr);
-		char strref[65000];
-		ReadNextLine(instr, strref, 65000);	// The reference string.
+		char strref[400000];
+		ReadNextLine(instr, strref, 400000);	// The reference string.
 		subStr=strchr(strref,'=')+1;
 		m_Header.m_Reference = subStr;
 	}
